pkg/didcomm/protocol/didexchange/persistence.go

```diff
@@ -94,13 +104,81 @@ func (c *ConnectionRecorder) GetInvitation(id string) (*Invitation, error) {
 	return result, nil
 }
 
-// GetConnectionRecord return connection record
+// GetConnectionRecord return connection record based on the connection ID
 func (c *ConnectionRecorder) GetConnectionRecord(connectionID string) (*ConnectionRecord, error) {
-	name, err := c.store.Get(connectionID)
+	k := connectionKeyPrefix(connectionID)
+	connRecordBytes, err := c.store.Get(k)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("get connection record:%w", err)
+	}
+	return prepareConnectionRecord(connRecordBytes)
+}
+
+// GetConnectionRecordByProtocolID return connection record via protocol ID
+func (c *ConnectionRecorder) GetConnectionRecordByProtocolID(protocolID string) (*ConnectionRecord, error) {
+	connectionIDBytes, err := c.store.Get(protocolID)
+	if err != nil {
+		return nil, fmt.Errorf("get connectionID by protocolID:%w", err)
+	}
+	// adding prefix for storing connection record
+	k := connectionKeyPrefix(string(connectionIDBytes))
+
+	connRecordBytes, err := c.store.Get(k)
+	if err != nil {
+		return nil, fmt.Errorf("get connection record by connectionID:%w", err)
+	}
+	return prepareConnectionRecord(connRecordBytes)
+}
+
+// saveConnectionRecord saves the connection record against the connection id  in the store
+func (c *ConnectionRecorder) saveConnectionRecord(record *ConnectionRecord) error {
+	k := connectionKeyPrefix(record.ConnectionID)
+	bytes, err := json.Marshal(record)
+	if err != nil {
+		return fmt.Errorf("save connection record:%w", err)
+	}
+	return c.store.Put(k, bytes)
+}
+
+// saveNewConnectionRecord saves newly created connection record against the connection id in the store
+// and it creates mapping from protocolID to connection ID
+func (c *ConnectionRecorder) saveNewConnectionRecord(record *ConnectionRecord) error {
+	if record.ThreadID == "" || record.ConnectionID == "" || record.Namespace == "" {
+		return fmt.Errorf("input parameters thid : %s and connectionId : %s namespace : %s cannot be empty",
+			record.ThreadID, record.ConnectionID, record.Namespace)
+	}
+	err := c.saveConnectionRecord(record)
+	if err != nil {
+		return err
+	}
+	return c.saveProtocolID(record.ThreadID, record.Namespace, record.ConnectionID)
+}
```

Comment:
there is no usage for `saveConnectionRecord` function, we do not need `saveConnectionRecord` function we can do everything in `saveNewConnectionRecord` e.g

```
func (c *ConnectionRecorder) saveNewConnectionRecord(record *ConnectionRecord) error {
	if record.ThreadID == "" || record.ConnectionID == "" || record.Namespace == "" {
		return fmt.Errorf("input parameters thid : %s and connectionId : %s namespace : %s cannot be empty",
			record.ThreadID, record.ConnectionID, record.Namespace)
	}

	bytes, err := json.Marshal(record)
	if err != nil {
		return fmt.Errorf("save connection record:%w", err)
	}

	err := c.store.Put(connectionKeyPrefix(record.ConnectionID), bytes)
	if err != nil {
		return err
	}

	return c.saveProtocolID(record.ThreadID, record.Namespace, record.ConnectionID)
}
```

Reply:
in addition, I would move validation to `isValid` function (need to create, should be part of ConnectionRecord) e.g `record.isValid()`

Reply:
the reason it is splitter because in saveNewConnectionRecord we are storing the record and the mapping of thread with the connection id.  This has to called only when new connection record is created. 

Other time we will just save the updated record in the store. using save connection record.

Reply:
ok